### ISSUE TITLE
URL instead of version number being saved in `package.json`

### ISSUE THREAD

**Issue by DullReferenceException** [npm/npm#15050]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [ ] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

The URL for packages being written to `package.json` instead of the version when using `--save` or `--save-dev` with `npm install`, when the dependency is already listed as a sub-dependency in `npm-shrinkwrap.json`. This is affecting the latest `3.10.x` version, and appears to have been introduced with version `3.10.4`. 

#### How can the CLI team reproduce the problem?

These are not the packages I'm using, but it easily reproduces the issue:

```shell
npm init
npm i --save-dev selenium-test-runner
npm shrinkwrap --dev
npm i --save-dev chai
```

The `devDependencies` in `package.json` now includes this line instead of a version number:

```
    "chai": "https://registry.npmjs.org/chai/-/chai-1.9.2.tgz",
```

If installing a package that is _not_ listed in the shrinkwrap file, then the version number is saved as expected.

### supporting information:

 - `npm -v` prints: `3.10.10`
 - `node -v` prints: `v6.7.0`
 - `npm config get registry` prints: `https://registry.npmjs.org/`
 - Windows, OS X/macOS, or Linux?: `OS X`
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by legodude17** [comment]
It is probably getting the `from` of `npm-shrinkwrap.json`. That doesn't seem like the correct behavior though. Any thoughts @zkat?

**Comment by iarna** [comment]
@legodude17 It used `resolved`, and while that's normal and intentional for _what gets installed_, it shouldn't be using that to save a version in your `package.json`. For your `package.json` it should probably be doing it's normal transform on `from` or `version`. That is, if the original origin of this package was the registry, then it should get your package prefix plus the version number. If the origin WAS a tarball then obviously it should remain using a tarball as the version.

**Comment by DullReferenceException** [comment]
I just noticed a similar bug, probably from the same root cause. The `from` properties in `npm-shrinkwrap.json` are also getting updated to the `resolved` values. For example, I just did an `npm i --save` for one of my dependencies, and `"from": "package@>=3.5.0 <4.0.0"` changed to `"from": "packageurl"`

**Comment by mnquintana** [comment]
I'm running into the same issue on npm 4.1.2, with a shrinkwrap file – is there any info I can provide that'd be helpful?

**Comment by xzyfer** [comment]
This appears to be happening with npm@5 - https://github.com/sass/node-sass/issues/1991

**Comment by ruchernchong** [comment]
This issue seemed to be not working on Ubuntu with npm v5.0.1. However, it is working on macOS. @xzyfer with reference to https://github.com/sass/node-sass/issues/1991#issuecomment-305391765

**Comment by sergejvoronov** [comment]
@xzyfer still persists on npm 5.0.1. [sass/node-sass#1991](https://github.com/sass/node-sass/issues/1991#issuecomment-305391765)

**Comment by almothafar** [comment]
Windows 10 here, npm 5.0.2 still issue persists, 

Have you checked the change log here: https://github.com/npm/npm/releases/tag/v5.0.0 ?

**Comment by karimsa** [comment]
On npm@`v5.1.0` now. Tested with Ubuntu 12.04 & macOS 10.12.5.

Issue showed up on Ubuntu but not on macOS.

**Comment by joshclow** [comment]
I hit this while digging into #17858. 
In my repro, I'm trying to install an internal dependency as a hosted tarball, so in the project's `package.json` there's an entry in dependencies that looks like 
`"myPackage": "https://foo.bar.fnord/myPackage-1.2.3.tgz"`

When the lock file gets built, `childVersion` in `shrinkwrap.js` on this entry falls into the middle case and returns `child.package._resolved || req.saveSpec || req.rawSpec`, and since `child.package._resolved` is the URL, then that's what ends up populating the corresponding version number in the lock file with that value and that ends up propagating down into the actual `package.json` file for the dependency if npm needs to install the package again (like if you rm -rf node_modules for example).

If I force the code there to just return `child.package.version`, then that ends up populating the lock file and everything from there on out is happy. 

That's not an actual _fix_ as such, I need to walk the code a bit more to know what the prod-ready version of this is, but it's clear at least this is a path that results in this behavior.

Once this has happened, `npm ls` and further `npm install`s will blow up in various ways, because any time something tries to use the `read-package-json` module, it will explode due to `semver` getting fed a URL.

**Comment by ruchernchong** [comment]
@joshclow what is your npm version?

**Comment by joshclow** [comment]
@ruchern This was with 5.5.1

I should also note that I'm getting in here via the creation of package-lock.json and not via an explicit use of `npm shrinkwrap`, but the end state is the same, with the URL replacing the version in the dependency's package.json.

**Comment by jcollum** [comment]
Mac OS 10.11.6

I'm also seeing this. 

```
✗ node --version && npm --version                     
v6.11.3
3.10.10
✗ npm i -S @company/local-config

```

Result in package.json: 

```
"@company/local-config": 
"http://ourlocalnpmregistry.com/@company%2flocal-config/-/local-config-0.0.8.tgz",
```

Pings etc 

```
✗ ping registry.npmjs.org
PING a.sni.fastly.net (151.101.192.162): 56 data bytes
64 bytes from 151.101.192.162: icmp_seq=0 ttl=56 time=65.409 ms
64 bytes from 151.101.192.162: icmp_seq=1 ttl=56 time=67.675 ms
64 bytes from 151.101.192.162: icmp_seq=2 ttl=56 time=68.655 ms
^C
--- a.sni.fastly.net ping statistics ---
5 packets transmitted, 5 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 65.409/67.499/69.778/1.627 ms
✗ npm view npm version

5.5.1

```

My .npmrc: 

```
✗ cat ~/.npmrc
strict-ssl=false
@company:registry=http://ourlocalnpmregistry.com
save-prefix=~
init.author.name=Justin Collum
init.author.email=justin.collum@company.com
registry=http://registry.npmjs.org/

```

The contributors to the project use a mix of NPM 3 and 5 so there is an `npm-shrinkwrap.json` file present but I am running npm 3. Could that be the issue?

**Comment by jcollum** [comment]
Mac OS 10.11.6

I'm also seeing this. 

```
✗ node --version && npm --version                     
v6.11.3
3.10.10
✗ npm i -S @company/local-config

```

Result in package.json: 

```
"@company/local-config": 
"http://ourlocalnpmregistry.com/@company%2flocal-config/-/local-config-0.0.8.tgz",
```

Pings etc 

```
✗ ping registry.npmjs.org
PING a.sni.fastly.net (151.101.192.162): 56 data bytes
64 bytes from 151.101.192.162: icmp_seq=0 ttl=56 time=65.409 ms
64 bytes from 151.101.192.162: icmp_seq=1 ttl=56 time=67.675 ms
64 bytes from 151.101.192.162: icmp_seq=2 ttl=56 time=68.655 ms
^C
--- a.sni.fastly.net ping statistics ---
5 packets transmitted, 5 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 65.409/67.499/69.778/1.627 ms
✗ npm view npm version

5.5.1

```

My .npmrc: 

```
✗ cat ~/.npmrc
strict-ssl=false
@company:registry=http://ourlocalnpmregistry.com
save-prefix=~
init.author.name=Justin Collum
init.author.email=justin.collum@company.com
registry=http://registry.npmjs.org/

```

The contributors to the project use a mix of NPM 3 and 5 so there is an `npm-shrinkwrap.json` file present but I am running npm 3. Could that be the issue?

**Comment by jcollum** [comment]
Mac OS 10.11.6

I'm also seeing this. 

```
✗ node --version && npm --version                     
v6.11.3
3.10.10
✗ npm i -S @company/local-config

```

Result in package.json: 

```
"@company/local-config": 
"http://ourlocalnpmregistry.com/@company%2flocal-config/-/local-config-0.0.8.tgz",
```

Pings etc 

```
✗ ping registry.npmjs.org
PING a.sni.fastly.net (151.101.192.162): 56 data bytes
64 bytes from 151.101.192.162: icmp_seq=0 ttl=56 time=65.409 ms
64 bytes from 151.101.192.162: icmp_seq=1 ttl=56 time=67.675 ms
64 bytes from 151.101.192.162: icmp_seq=2 ttl=56 time=68.655 ms
^C
--- a.sni.fastly.net ping statistics ---
5 packets transmitted, 5 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 65.409/67.499/69.778/1.627 ms
✗ npm view npm version

5.5.1

```

My .npmrc: 

```
✗ cat ~/.npmrc
strict-ssl=false
@company:registry=http://ourlocalnpmregistry.com
save-prefix=~
init.author.name=Justin Collum
init.author.email=justin.collum@company.com
registry=http://registry.npmjs.org/

```

The contributors to the project use a mix of NPM 3 and 5 so there is an `npm-shrinkwrap.json` file present but I am running npm 3. Could that be the issue?

**Comment by jcollum** [comment]
I confirmed this is a bug involving `npm-shrinkwrap.json`. 

1. Removed npm-shrinkwrap.json
1. installed `@company/local-config`
1. Result: `"@company/local-config": "0.0.8",`
1. Restored npm-shrinkwrap.json
1.  installed `@company/local-config`
1. Result: `"@company/local-config": 
"http://ourlocalnpmregistry.com/@company%2flocal-config/-/local-config-0.0.8.tgz",`

Yes, `@company/local-config` is present in `npm-shrinkwrap.json`. I also saw it happen with `moment` library. 

This appears to be a result of running NPM 5 and NPM 3 in a team that is using Node 6 to build a project. But no one would do that, that would be wacky! Oh, but we are 🤕
